clippy_utils/src/lib.rs

```diff
@@ -379,6 +379,39 @@ pub fn match_qpath(path: &QPath<'_>, segments: &[&str]) -> bool {
     }
 }
 
+/// Resolves the path to a `DefId` and checks if it matches the given path.
+pub fn match_qpath_res(cx: &LateContext<'_>, path: &QPath<'_>, hir_id: HirId, segments: &[&str]) -> bool {
+    cx.qpath_res(path, hir_id)
+        .opt_def_id()
+        .map_or(false, |id| match_def_path(cx, id, segments))
+}
+
+/// If the expression is a path, resolves it to a `DefId` and checks if it matches the given path.
+pub fn match_expr_path_res(cx: &LateContext<'_>, expr: &Expr<'_>, segments: &[&str]) -> bool {
+    if let ExprKind::Path(p) = &expr.kind {
+        match_qpath_res(cx, p, expr.hir_id, segments)
+    } else {
+        false
+    }
+}
+
+/// Resolves the path to a `DefId` and returns the index of the first matching path, if any.
+pub fn match_any_qpath_res(cx: &LateContext<'_>, path: &QPath<'_>, hir_id: HirId, paths: &[&[&str]]) -> Option<usize> {
+    cx.qpath_res(path, hir_id)
+        .opt_def_id()
+        .and_then(|id| match_any_def_paths(cx, id, paths))
+}
+
+/// If the expression is a path, resolves it to a `DefId` and returns the index of the first
+/// matching path, if any.
+pub fn match_any_expr_path_res(cx: &LateContext<'_>, expr: &Expr<'_>, paths: &[&[&str]]) -> Option<usize> {
+    if let ExprKind::Path(p) = &expr.kind {
+        match_any_qpath_res(cx, p, expr.hir_id, paths)
+    } else {
+        None
+    }
+}
```

Comment:
I don't think we should have a util for every combination of Expr/QPath/DefId and 1 or N paths. Perhaps just add one util like this?

```rust
pub fn path_res_def_id(cx, expr) -> Option<DefId>;
```

Reply:
My argument for them is just the sheer frequency they would be used. Both `match_qpath_res` and `match_expr_path_res` would probably have over a hundred usages each if they were used everywhere they could be. The ones for multiple paths are a little less defensible.

Reply:
Type relative paths still need to be resolved for trait functions so `qpath_res` is needed here.

Reply:
There is a lot of code which does something like this:

```rust
if_chain {
    if let ExprKind::Call(path_expr, [arg1, arg2, ...]) = expr.kind;
    if let ExprKind::Path(p) = &path_expr.kind;
    if let Some(id) = cx.qpath_res(p, path_expr.hir_id).opt_def_id();
    if match_def_path(cx, id, &SOME_PATH);
    then {
        do_something();
    }
}
```

Ideally something like `if let Some(args) = match_call(cx, expr, &SOME_PATH)` would work. A function would be sub-optimal as the argument list can't be pattern matched before doing all the work to resolve the path.

Matching pattern paths is fairly rare, `TupleStruct` and `Struct` are usually also being checked for at the same time. I doubt something like `match_pat_path_res` would see much use. I assume the plan is ultimately to use diagnostic items for everything. That just leaves `match_expr_path_diagnostic_item` (and maybe one for `LangItem`s).

There are also a few cases where it could be used as a guard clause for match arms. e.g.

```rust
match expr.kind {
    ExprKind::Call(func, [arg]) if match_expr_path_res(cx, func, &SOME_PATH) => do_something(),
    // Some other arms
    _ => None,
}
```

With something like `expr_path_def_id` that would look like:

```rust
match expr.kind {
    ExprKind::Call(func, [arg])
        if expr_path_def_id(cx, func)
            .map_or(false, |id| match_def_path(cx, id, &SOME_PATH)) =>
    {
        do_something()
    }
    // Some other arms
    _ => None,
}
```

It's not a common occurrence, but would be nice when it comes up.

Reply:
Yeah I can get behind `match_expr_path_res`. For the name...`is_expr_path_def_path`? Or `is_path_to_def_path`? I kinda like the word "to" since it communicates path resolution. But I don't have a strong opinion on this one.

Reply:
So maybe `is_qpath_to_def_path` and `is_expr_path_to_def_path`. It sounds weird to me just because I read it as `is_path_to_path`. It does get the point across better though. It also reads well with `diagnostic_item` and `lang_item`.

Reply:
Unrelated question, is there a reason for `is_lang_ctor` rather than just `is_lang_item`? Is is possible for the path to resolve to the variant and not the constructor that isn't already disambiguated by where the path came from?

Reply:
When creating an emum value like `Some(..)` or `None`, the path will resolve to the ctor.

Reply:
I mean why is it restricted to just constructors?

Reply:
Because (from what I found at least) it seems to cover almost all lang item use cases.

Reply:
Ok. I've left `match_any_def_paths` just because there is a performance over the obvious way of using `Iterator::any`. All the other multi-path functions have bee removed.